node/impl/client/client.go

```diff
@@ -495,84 +484,120 @@ func (a *API) makeRetrievalQuery(ctx context.Context, rp rm.RetrievalPeer, paylo
+		}
+
+		carFile = ref.Path
+		root = hd.Roots[0]
+	} else {
+		carFile, err = a.imgr().NewTempFile(id)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to create temp CARv2 file: %w", err)
+		}
+		// make sure to remove the CARv2 file if anything goes wrong from here on.
+		defer func() {
+			if finalErr != nil {
+				_ = os.Remove(carFile)
+			}
+		}()
+
+		root, err = a.doImport(ctx, ref.Path, carFile)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to import normal file to CARv2: %w", err)
+		}
 	}
 
-	if err := a.imgr().AddLabel(id, importmgr.LFileName, ref.Path); err != nil {
+	if err := a.imgr().AddLabel(id, imports.LSource, "import"); err != nil {
 		return nil, err
 	}
-
-	nd, err := a.clientImport(ctx, ref, st)
-	if err != nil {
+	if err := a.imgr().AddLabel(id, imports.LFileName, ref.Path); err != nil {
 		return nil, err
 	}
-
-	if err := a.imgr().AddLabel(id, importmgr.LRootCid, nd.String()); err != nil {
+	if err := a.imgr().AddLabel(id, imports.LCARPath, carFile); err != nil {
+		return nil, err
+	}
+	if err := a.imgr().AddLabel(id, imports.LRootCid, root.String()); err != nil {
 		return nil, err
 	}
 
 	return &api.ImportRes{
-		Root:     nd,
+		Root:     root,
 		ImportID: id,
 	}, nil
 }
 
-func (a *API) ClientRemoveImport(ctx context.Context, importID multistore.StoreID) error {
-	return a.imgr().Remove(importID)
-}
-
-func (a *API) ClientImportLocal(ctx context.Context, f io.Reader) (cid.Cid, error) {
-	file := files.NewReaderFile(f)
-
-	id, st, err := a.imgr().NewStore()
+func (a *API) ClientRemoveImport(ctx context.Context, importID imports.ID) error {
+	info, err := a.imgr().Info(importID)
 	if err != nil {
-		return cid.Undef, err
+		return xerrors.Errorf("failed to fetch import info: %w", err)
 	}
-	if err := a.imgr().AddLabel(id, "source", "import-local"); err != nil {
-		return cid.Cid{}, err
+
+	// remove the CARv2 file if we've created one.
+	if path := info.Labels[imports.LCARPath]; path != "" {
+		_ = os.Remove(path)
 	}
 
-	bufferedDS := ipld.NewBufferedDAG(ctx, st.DAG)
+	return a.imgr().Remove(importID)
+}
 
-	prefix, err := merkledag.PrefixForCidVersion(1)
+func (a *API) ClientImportLocal(ctx context.Context, r io.Reader) (cid.Cid, error) {
+	// write payload to temp file
+	tmpPath, err := a.imgr().NewTempFile(imports.ID(rand.Uint64()))
 	if err != nil {
 		return cid.Undef, err
 	}
-	prefix.MhType = DefaultHashFunction
-
-	params := ihelper.DagBuilderParams{
-		Maxlinks:  build.UnixfsLinksPerLevel,
-		RawLeaves: true,
-		CidBuilder: cidutil.InlineBuilder{
-			Builder: prefix,
-			Limit:   126,
-		},
-		Dagserv: bufferedDS,
-	}
-
-	db, err := params.New(chunker.NewSizeSplitter(file, int64(build.UnixfsChunkSize)))
+	defer os.Remove(tmpPath) //nolint:errcheck
+	tmpF, err := os.Open(tmpPath)
 	if err != nil {
 		return cid.Undef, err
 	}
-	nd, err := balanced.Layout(db)
-	if err != nil {
+	defer tmpF.Close() //nolint:errcheck
+	if _, err := io.Copy(tmpF, r); err != nil {
```

Comment:
This is a performance regression. We just replaced direct UnixFS chunking into a "heavy" CAR file (containing leaf data) with two files (the temp file where we wrote the upload, and the position CAR file). I don't think that's right.

Reply:
I think we should import straight into a heavy CAR.

Reply:
Fixed.